### PR TITLE
Arithmetic overflow error for NUMERIC datatype

### PR DESCRIPTION
**Pull request by bsatarupa**
### Problem description

Existing bbf code throws overflow error for multiplication or division of numeric operands whenever the resultant calculated precision > TDS_MAX_NUMERIC_PRECISION(i.e., 38) and resultant calculated integral part(i.e., precision - scale) > 32 and resultant  calculated scale > 6

Ideally, the above scenario may and *may not* cause overflow and that decision is made later based on the actual numeric datalength (which is already handled as part of resolving BABEL-3179).

We have fixed the above mentioned issue, by rounding scale to 6 and precision to TDS_MAX_NUM_PRECISION.(mimicking sql server behaviour: https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/t-sql/data-types/precision-scale-and-length-transact-sql.md) and deferred the overflow checking decision to a later phase when integral part of actual data can't be accommodated in 32 digits (i.e., entire numeric result can't be accommodated in TDS_MAX_NUM_PRECISION digits since scale is already set to 6).

#### Test scenarios covered

o Use case based:Original testcase that was causing arithmetic overflow but currently returning numeric value with scale=6.
o Boundary conditions: Testcases satisfying/ not satisfying the above condition.
o Arbitrary inputs : Done
o Null, Negative test scenarios: N/A
o Minor version upgrade tests, Major version upgrade tests : Done
o Performance tests, Tooling impact, Client tests: N/A

Task: BABEL-3345
Signed-off-by: Satarupa Biswas <satarupb@amazon.com>

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by rishabhtanwar29 — DISMISSED** [review]

File: contrib/babelfishpg_tds/src/backend/tds/tdsresponse.c [hunk 1]
```diff
@@ -564,18 +564,31 @@ resolve_numeric_typmod_from_exp(Node *expr)
...
-			     * and (multiplcation having no aggregate operand)
-			     */
-			    int delta = precision - TDS_MAX_NUM_PRECISION;
-			    precision = TDS_MAX_NUM_PRECISION;
-			    scale = Max(scale - delta, 0);
+				if (precision - scale > 32 && scale > 6)
+				{
+					/*
+					 * Result might be rounded to 6 decimal places or the overflow error
+					 * will be thrown if the integral part can't fit into 32 digits.
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tds/src/backend/tds/tdsresponse.c:573]: where will we detect this overflow if intergral part can't fit in 38 places? Also, why we are only limiting scale to 6? what if intergral part required more than 32 palces

File: contrib/babelfishpg_tds/src/backend/tds/tdsresponse.c [hunk 2]
```diff
@@ -564,18 +564,31 @@ resolve_numeric_typmod_from_exp(Node *expr)
...
+					/*
+					 * scale adjustment by delta is only applicable for division
+			     		 * and (multiplcation having no aggregate operand)
+			     		 */
+			    		int delta = precision - TDS_MAX_NUM_PRECISION;
+			    		precision = TDS_MAX_NUM_PRECISION;
+			    		scale = Max(scale - delta, 0);
+				}
+				/*
+				 * Control reaching here for only arithmetic overflow cases
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tds/src/backend/tds/tdsresponse.c:589]: Have we added tests to verify both if and else if and else case for the change?

**Review by surendravishnoi — COMMENTED** [review]
Please add template to PR description. Also, add comments on tests to explain what we are verifying.

File: contrib/babelfishpg_tds/src/backend/tds/tdsresponse.c (hunk 2 above)
Comment (bsatarupa) [reply contrib/babelfishpg_tds/src/backend/tds/tdsresponse.c:589]: Testcases added for all conditional statements.

File: contrib/babelfishpg_tds/src/backend/tds/tdsresponse.c (hunk 1 above)
Comment (bsatarupa) [reply contrib/babelfishpg_tds/src/backend/tds/tdsresponse.c:573]: Q: where will we detect this overflow if intergral part can't fit in 38 places?
A: When integral part can't be handled in 38 places, we throw overflow error based on actual numeric data length. [This part was added in BABEL-3179[https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/40 and https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/495]

Q: Also, why we are only limiting scale to 6? what if intergral part required more than 32 places?
A. We are following sql server behaviour, where if (precision - scale > 32 && scale > 6), we set scale to 6 and if integral part can't be accommodated in rest of 32 digits, overflow error is triggered from TDS side.
Followed behaviour: https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/t-sql/data-types/precision-scale-and-length-transact-sql.md

File: test/JDBC/expected/TestDecimal-vu-prepare.out [hunk 3]
```diff
@@ -404,3 +404,14 @@ CREATE TABLE testdecimal_overflow_test5 (id integer PRIMARY KEY, amount1 decimal
 INSERT INTO testdecimal_overflow_test5 VALUES (1, 9999.99, 999.999, 999.12345678), (2, 9999.99, 999.999, 999.12345678), (3, 9999.99, 999.999, 999.12345678), (4, 9999.99, 999.999, 999.12345678), (5, 9999.99, 999.999, 999.12345678), (6, 9999.99, 999.999, 999.12345678), (7, 9999.99, 999.999, 999.12345678), (8, 9999.99, 999.999, 999.12345678), (9, 9999.99, 999.999, 999.12345678), (10,9999.99, 999.999, 999.12345678), (11, 9999.99, 999.999, 999.12345678), (12, 9999.9, 999.999, 999.12345678), (13, 9999.99, 999.999, 999.12345678), (14,9999.99, 999.999, 999.12345678), (15, 9999.99, 999.999, 999.12345678), (16, 9999.99, 999.999, 999.12345678), (17, 9999.99, 999.999, 999.12345678), (18, 9999.99, 999.999, 999.12345678), (19, 9999.99, 999.999, 999.12345678), (20,9999.99, 999.999, 999.12345678), (21, 9999.99, 999.999, 999.12345678), (22,9999.99, 999.999, 999.12345678);
 ~~ROW COUNT: 22~~
 
+
+#BABEL-3345
+create table test_dec (int1 decimal(24,6),int2 decimal(24,6),rate decimal(24,6), type varchar(1000));
```
Comment (rishabhtanwar29) [inline test/JDBC/expected/TestDecimal-vu-prepare.out]: We should follow the object naming convention here (object name should start with filename).
i.e. : testdecimal_vu_prepare_*.
Let's follow this convention for all the test files.

Comment (bsatarupa) [reply test/JDBC/expected/TestDecimal-vu-prepare.out]: done

**Review by macwoody — APPROVED** [review]

**Review by rishabhtanwar29 — APPROVED** [review]

**Review by forestkeeper — APPROVED** [review]
LGTM
